internal/secrets/secrets.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package secrets provides the interface for interacting with Secret Manager
+package secrets
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	secretmanager "cloud.google.com/go/secretmanager/apiv1"
+	"cloud.google.com/go/secretmanager/apiv1/secretmanagerpb"
+	"github.com/googleapis/gax-go/v2"
+	"github.com/googleapis/gax-go/v2/apierror"
+	"google.golang.org/grpc/codes"
+)
+
+// A provider struct that holds context for interacting with Secret Manager
+type Client struct {
+	// ctx is the context used to make Secret Manager requests
+	ctx context.Context
+
+	// secretsClient is used to fetch secrets from Secret Manager
+	secretsClient SecretsClient
+
+	// project is the GCP project in which to look up secrets
+	project string
+
+	// secretCache is an in-memroy cache of secrets we've already fetched
+	secretCache map[string]string
+}
+
+// An interface used for testing. A secretmanager.Client implements
+// this interface
+type SecretsClient interface {
+	AccessSecretVersion(ctx context.Context, req *secretmanagerpb.AccessSecretVersionRequest, opts ...gax.CallOption) (*secretmanagerpb.AccessSecretVersionResponse, error)
+}
+
+func New(ctx context.Context, project string, secretsClient SecretsClient) (*Client, error) {
+	var client SecretsClient
+	if secretsClient != nil {
+		client = secretsClient
+	} else {
+		secretsClient, err := secretmanager.NewClient(ctx)
+		if err != nil {
+			return nil, err
+		}
+		client = secretsClient
+	}
+	return &Client{
+		ctx:           ctx,
+		secretsClient: client,
+		project:       project,
+		secretCache:   make(map[string]string),
+	}, nil
+}
+
+// Given a SecretsProvider context, fetch the latest version of a secret
+// as a string. This function uses a read-through cache to reduce the
+// number of API calls needed. This method assumes the secret payload is
+// a UTF-8 string.
+func (s *Client) GetSecretManagerValue(secretName string) (string, bool, error) {
```

Comment:
I think you can simplify this whole package to this:

https://go.dev/play/p/5F0Z_5p2Eux

https://github.com/golang/pkgsite/tree/master/internal/secrets has an example implementation

Reply:
Done. Not that there's not a good way to unit test this code as it's making a network request and can't easily be mocked (unless we dependency inject a transport). Note that the example implementation's test is an integration test.